### DISCUSSION TITLE
Increase number of retries or timeout on Android

### DISCUSSION THREAD

**Discussion by vicentiu2** [strongswan/strongswan#1565] [Q&A]
Hello, when I try to connect on Android, it triggers 2FA and it's usually quite slow and it takes longer than the 3 retries that the app does.
By the time the 2FA arrives for approval, the app has already exhausted its 3 retries.
I have to try over and over again until rarely a 2FA approval request comes in before the app's 3 retries are exhausted.

How can the number of retries be increased or the timeout for each retry?

Extending the authentication retries/timeout by another 20 seconds or so would suffice.

thank you

**Comment by tobiasbrunner** [comment]
> it triggers 2FA and it's usually quite slow and it takes longer than the 3 retries that the app does

What kind of 2FA authentication? How does it work exactly?

> How can the number of retries be increased or the timeout for each retry?

That's currently not possible without building your own version of the app. I guess we could add an option to increase the number of retries in a future update.

**Reply by vicentiu2** [reply]
After clicking Connect, a 2FA is trigerred through the Microsoft Authenticator app which comes back for approval through a push notification with Approve/Deny buttons.

Once I click Approve, the VPN authentication succeeds.

Yes, having an option to increase the number of retries would be great.

thanks

**Reply by tobiasbrunner** [reply]
An alternative to making this configurable that I'm currently considering, is using different retransmission settings for `IKE_SA_INIT` (i.e. the initial message). Then the app could use the current settings for that exchange and abort relatively quickly (currently ~14 seconds) if the server is not reachable, but for later exchanges we could allow more time and configure additional retransmits (e.g. 6 instead of 3, which would increase the overall timeout to 48 seconds). That would not require users to configure anything and the overall behavior should still  be about the same, except that users with setups like yours have more time. (_Edit: Fixed the times._)

**Reply by vicentiu2** [reply]
49 seconds would suffice. My experience is Strongswan times out in 15-18 seconds and the 2FA approval notification comes 22-35 seconds after trying to connect.

**Comment by vicentiu2** [comment]
Never mind. I found out my phone was really laggy due to an app that was hogging the cpu. Once the app was removed, 2FA notifications always arrive in a timely manner.

**Reply by tobiasbrunner** [reply]
Great, thanks for the update.
